File: Solution 2.cpp
class Solution 
{
public:
    int minimumTimeRequired(vector<int>& jobs, int k) 
    {
        int n = jobs.size();
        vector<int> state_jobsum((1<<n), 0);
        for (int i = 0; i < n; i ++)
            for (int j = 0; j < (1 << i); j ++)
                state_jobsum[(1<<i) + j] = jobs[i] + state_jobsum[j];
            
        vector<vector<int>> dp(k, vector<int>((1<<n), 0));
        for (int state = 0; state < (1<<n); state ++)
            dp[0][state] = state_jobsum[state];
        
        for (int i = 1; i < k; i ++)
        {
            for (int state = 0; state < (1<<n); state ++)
            {
                int minval = INT_MAX;
                int x = state;
                while(x != 0)
                {
                    int cur = max(dp[i-1][state-x], state_jobsum[x]);
                    minval = min(minval, cur);
                    x = (x - 1) & state;
                }
                dp[i][state] = minval;
            }
        }

        return dp[k-1][(1<<n) - 1];
    }
};

File: Solution 2.py
class Solution:
    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        jobLen = len(jobs)
        state_jobsum = [0 for _ in range(1 << jobLen)]
        for i in range(jobLen):
            for j in range(1 << i):
                state_jobsum[(1 << i) + j] = jobs[i] + state_jobsum[j]

        dp = [[0 for _ in range(1 << jobLen)] for _ in range(k)]
        for i in range(1 << jobLen):
            dp[0][i] = state_jobsum[i]

        for i in range(1, k):
            for j in range(1 << jobLen):
                minval = float('inf')
                x = j
                while x != 0:
                    cur = max(state_jobsum[x], dp[i-1][j-x])
                    minval = min(minval, cur)
                    x = (x - 1) & j
                dp[i][j] = minval
        
        return dp[k-1][(1 << jobLen) - 1]
